### ISSUE TITLE
ScintillaNet examples for VB.NET

### ISSUE THREAD

**Issue by byman64** [jacobslusser/ScintillaNET#27]
1st of all thanks for sharing ScintillaNET.
it's a great idea

I am searching an example in vb.net to understand better how to create my own basic editor.
I tried to use ScintillaNet 2,  with less difficults because I found and example.
I need to integrate in my project an SQL editor with basic and strong feature like:
- Highlight code
- Autocomplete:
a) SQL keywords
b) a couple of my own stuff like list databases, list of tables, list of column
c) I would like add the all  words of current text, maybe I have to do it myself or just active a flag because it is a feature already embedded in scintilla?
d) If it is possible add  for each keyword, a description and a prefix. Prefix usefull to group my own keywords. For example if I write #db I would like see the list of databases and when I select one of if  I don't include  the prefix #db in my text.
- Maybe the "call tips" can help me refering the above request? I would like an example.

Where i can find an example?
grazie mille
Antonino Migliore

**Comment by JohnyMac** [comment]
This website can help you a bit: http://converter.telerik.com/

Good Luck, And if you need any assistance... Reply here.

**Comment by byman64** [comment]
Thanks for your link, I already know that, my main trouble is not the language c...

For instance, I just create a new test project and I added the following code in the form1.
I was waiting for "Automatic Syntax Highlighting" for sql language. But if I seems not working.

Thanks in advance

' CODE
Imports ScintillaNET
Public Class Form1

    Private Sub Form1_Load(sender As Object, e As System.EventArgs) Handles Me.Load
        Scintilla1.Lexer = Lexer.sql
      End Sub
End Class

**Comment by jacobslusser** [comment]
@byman64, 

Your comments suggest to me that you've aware of the wiki article "[Automatic Syntax Highlighting](https://github.com/jacobslusser/ScintillaNET/wiki/Automatic-Syntax-Highlighting)", but your code suggests to me that you have not read it completely. To get syntax highlighting to work requires more than just setting the `Lexer` property.

Try these articles and let me know if you still have questions:
- [Automatic Syntax Highlighting](https://github.com/jacobslusser/ScintillaNET/wiki/Automatic-Syntax-Highlighting)
- [Basic Autocompletion](https://github.com/jacobslusser/ScintillaNET/wiki/Basic-Autocompletion)


Jacob

**Comment by byman64** [comment]
Thanks for you help of course you are right. 

Please, confirm if I am in the right way

SYNTAX HIGHLIGHTING
========================================================================
- style definition are mandatory (now Syntax Highlighting is working fine)
- Scintilla1.SetKeywords(0, "abs absolute ..."). I understood the index 0, 1, 2..but there is not a specific name for the language .. the only one is on
 "Scintilla1.Lexer = Lexer.Sql" 

Here my vb.net code for Syntax Highlighting, it could be useful for others
// CODE BEGIN
Imports ScintillaNET

Public Class Form1

    Private Sub Form1_Load(sender As Object, e As System.EventArgs) Handles Me.Load
        Scintilla1.Lexer = Lexer.Sql
        'Scintilla1.Text = Scintilla1.DescribeKeywordSets(Lexer.Sql)

        ' Configuring the default style with properties
        ' we have common to every lexer style saves time.
        Scintilla1.StyleResetDefault()

        Scintilla1.Styles(Style.Default).Font = "Consolas"
        Scintilla1.Styles(Style.Default).Size = 10

        Scintilla1.StyleClearAll()

        ' Configure the Sql lexer styles
        Scintilla1.Styles(Style.Sql.Default).ForeColor = Color.Silver
        Scintilla1.Styles(Style.Sql.Comment).ForeColor = Color.FromArgb(0, 128, 0)
        ' Green
        Scintilla1.Styles(Style.Sql.CommentLine).ForeColor = Color.FromArgb(0, 128, 0)
        ' Green
        Scintilla1.Styles(Style.Sql.CommentLineDoc).ForeColor = Color.FromArgb(128, 128, 128)
        ' Gray
        Scintilla1.Styles(Style.Sql.Number).ForeColor = Color.Olive
        Scintilla1.Styles(Style.Sql.Word).ForeColor = Color.Blue
        Scintilla1.Styles(Style.Sql.Word2).ForeColor = Color.Blue
        Scintilla1.Styles(Style.Sql.[String]).ForeColor = Color.FromArgb(163, 21, 21)
        ' Red
        Scintilla1.Styles(Style.Sql.Character).ForeColor = Color.FromArgb(163, 21, 21)

        '  Full keyword removed, pls copy them from %appdata%\notepad++\lang.xml
        Scintilla1.SetKeywords(0, "abs absolute access...write year zone") 
        Scintilla1.Text = Scintilla1.DescribeKeywordSets()  ' debug line: write the style in the editor scintilla
    End Sub
// CODE END


AUTO COMPLETATION
==============================================================
It starting works if I write a keyword.
QUESTIONS
- I would like press Ctrl-Space to show the list
In previous version of Scintilla I didnt do anything: Ctrl-Space=List of keywords, Ctrl-F=Find and more.
Do I write some codes to do it? 

- To manage keywords composed with 2 words I think have to change the list separator like in the previous version. for example "ORDER BY". Which is your suggest?

Here the only code I have for autocompletation, it seems similar to the previous version of ScintillaNET.

// CODE BEGIN
Private Sub Scintilla1_CharAdded1(sender As Object, e As ScintillaNET.CharAddedEventArgs) Handles Scintilla1.CharAdded
        ' Find the word start
        Dim currentPos = Scintilla1.CurrentPosition
        Dim wordStartPos = Scintilla1.WordStartPosition(currentPos, True)

        ' Display the autocompletion list
        Dim lenEntered = currentPos - wordStartPos
        If lenEntered > 0 Then
            Scintilla1.AutoCShow(lenEntered, "abs absolute ... write year zone")
        End If
    End Sub
// CODE END

thanks again

**Comment by jacobslusser** [comment]
> Scintilla1.SetKeywords(0, "abs absolute ..."). I understood the index 0, 1, 2..but there is not a specific name for the language .. the only one is on "Scintilla1.Lexer = Lexer.Sql"

I'm not sure I understand this statement. Are you indicating a problem? Please clarify.

> I would like press Ctrl-Space to show the list
In the previous version of ScintillaNET I didn't write any line of code.
Ctrl-Space=List of keywords, Ctrl-F=Find and more.
Do I write some codes to do basic operations like Find, Replace, List of Kewords?

You are correct that the new version of ScintillaNET doesn't provide all the added support for language features like `CTRL+SPACE` that the old version did. The reasons why have already been discussed at length in the project Issues so I won't repeat them. To support `CTRL+SPACE` you'll need to handle the necessary key-press events yourself and call `AutoCShow`. Same is true for `CTRL+F`.

Technically speaking, the SQL keywords "ORDER" and "BY are not a single keyword/token. While it is true that using one without the other is a syntax error (these are sometimes called "contextual keywords") they are in fact separate keywords. To try and group them together would be difficult because any number of whitespace combinations can be used between them. i.e. the following are all valid SQL syntax:

```
ORDER BY
ORDER              BY
ORDER
BY
```

In other words, treat them separately in your keyword list.

**Comment by byman64** [comment]
...
> Scintilla1.SetKeywords(0, "abs absolute ...").
It's not an issue just a comment,
Scintilla1.SetKeywords it's not refered to any language. I expected something specific to the language, Something like Scintilla1.sql.SetKeywords.

- Ctrl-Sapce handled, so now I have the keyword list.

- Regarding find and replace which is the equivalent of old method .FindReplace.ShowFind()  ?

**Comment by jacobslusser** [comment]
There is no find (and replace) dialog built-in to ScintillaNET v3. See this search of the project Issues for more explanation:

[https://github.com/jacobslusser/ScintillaNET/issues?q=dialog](https://github.com/jacobslusser/ScintillaNET/issues?q=dialog)
